### ISSUE TITLE
Start-Transcript output is hard to parse

### ISSUE THREAD

**Issue by petemounce** [PowerShell/PowerShell#8720]
(This is feedback on the `Start-Transcript` cmdlet itself, the docs are great)

The output looks like

```
**********************
Windows PowerShell transcript start
Start time: 20190123080857
Username: BK-WINDOWS-V2-1\petermounce
RunAs User: BK-WINDOWS-V2-1\petermounce
Machine: BK-WINDOWS-V2-1 (Microsoft Windows NT 10.0.14393.0)
Host Application: C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
Process ID: 7264
PSVersion: 5.1.14393.2636
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.14393.2636
BuildVersion: 10.0.14393.2636
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************
Transcript started, output file is c:/blah\PowerShell_transcript.BK-WINDOWS-V2-1.0zNarV8d.20190123080857.txt
Logging a thing...
Done.
**********************
Command start time: 20190123080945
**********************
```
This is fine for human-reading, but very hard to parse for machines. For more context - I use Elastic's filebeat to hoover log files (and winlogbeat for event logs) into Elasticsearch. This enriches every log line with contextual information (like the things in the context-start section) so that _every log line_ gets turned into an event that is saved with that information. This makes my logs queryable in really useful ways within Kibana.

Would it be possible to do one of the following:
* Add a flag like `-OutputFormat=human|json` - human = as now, `json` = JSON blob per line (_not_ one single JSON object with an array of lines), where JSON blob contains the `message` from that line of transcript and also one field per item of contextual data
* Add a flag like `-LineTemplate` allowing the contextual information to be placed into the output according to how the user wants?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 61f8b565-4cc3-6a77-94e6-15b97ebcbca7
* Version Independent ID: afe03c1f-3dbf-62f7-722a-d845777cadbd
* Content: [Start-Transcript (Microsoft.PowerShell.Host)](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.host/start-transcript?view=powershell-6)
* Content Source: [reference/6/Microsoft.PowerShell.Host/Start-Transcript.md](https://github.com/PowerShell/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Host/Start-Transcript.md)
* Product: **powershell**
* GitHub Login: @SteveL-MSFT
* Microsoft Alias: **slee**

**Comment by lukexjeremy** [comment]
🤔 Could you provide a hypothetical example of a log produced with `-OutputFormat JSON`?

**Comment by petemounce** [comment]
@lukexjeremy I would find output of this form immediately consumable by tools such as filebeat. One blob per line, and the example below is a translation of the human-readable in the original ticket.

```
{"transcript":"Windows PowerShell transcript start","startTimeUtc":20190123080857,"username":"BK-WINDOWS-V2-1\\petermounce","runAsUser":"BK-WINDOWS-V2-1\\petermounce","machine":"BK-WINDOWS-V2-1 (Microsoft Windows NT 10.0.14393.0)","hostApplication":"C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe","processId":7264,"psVersion":"5.1.14393.2636","psEdition":"Desktop","psCompatibleVersions":["1.0","2.0","3.0","4.0","5.0","5.1.14393.2636"],"buildVersion":"10.0.14393.2636","clrVersion":"4.0.30319.42000","wsManStackVersion":"3.0","psRemotingProtocolVersion":"2.3","serializationVersion":"1.1.0.1","stdout":"Logging a thing...","stderr":""}
{"transcript":"Windows PowerShell transcript start","startTimeUtc":20190123080857,"username":"BK-WINDOWS-V2-1\\petermounce","runAsUser":"BK-WINDOWS-V2-1\\petermounce","machine":"BK-WINDOWS-V2-1 (Microsoft Windows NT 10.0.14393.0)","hostApplication":"C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe","processId":7264,"psVersion":"5.1.14393.2636","psEdition":"Desktop","psCompatibleVersions":["1.0","2.0","3.0","4.0","5.0","5.1.14393.2636"],"buildVersion":"10.0.14393.2636","clrVersion":"4.0.30319.42000","wsManStackVersion":"3.0","psRemotingProtocolVersion":"2.3","serializationVersion":"1.1.0.1","stdout":"Done.","stderr":""}
```

For readability within this ticket:

```json
{
    "transcript": "Windows PowerShell transcript start",
    "startTimeUtc": 20190123080857,
    "username": "BK-WINDOWS-V2-1\\petermounce",
    "runAsUser": "BK-WINDOWS-V2-1\\petermounce",
    "machine": "BK-WINDOWS-V2-1 (Microsoft Windows NT 10.0.14393.0)",
    "hostApplication": "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
    "processId": 7264,
    "psVersion": "5.1.14393.2636",
    "psEdition": "Desktop",
    "psCompatibleVersions": ["1.0", "2.0", "3.0", "4.0", "5.0", "5.1.14393.2636"],
    "buildVersion": "10.0.14393.2636",
    "clrVersion": "4.0.30319.42000",
    "wsManStackVersion": "3.0",
    "psRemotingProtocolVersion": "2.3",
    "serializationVersion": "1.1.0.1",
    "stdout": "Logging a thing...",
    "stderr": ""
}
{
    "transcript": "Windows PowerShell transcript start",
    "startTimeUtc": 20190123080857,
    "username": "BK-WINDOWS-V2-1\\petermounce",
    "runAsUser": "BK-WINDOWS-V2-1\\petermounce",
    "machine": "BK-WINDOWS-V2-1 (Microsoft Windows NT 10.0.14393.0)",
    "hostApplication": "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
    "processId": 7264,
    "psVersion": "5.1.14393.2636",
    "psEdition": "Desktop",
    "psCompatibleVersions": ["1.0", "2.0", "3.0", "4.0", "5.0", "5.1.14393.2636"],
    "buildVersion": "10.0.14393.2636",
    "clrVersion": "4.0.30319.42000",
    "wsManStackVersion": "3.0",
    "psRemotingProtocolVersion": "2.3",
    "serializationVersion": "1.1.0.1",
    "stdout": "Done.",
    "stderr": ""
}
```

**Comment by iSazonov** [comment]
I see huge duplications in the proposed format.

**Comment by vexx32** [comment]
Yeah, that's my main concern there as well. The information is good to have, but emitting it for _every line_ of logging will create performance concerns eventually.

The difficulty here is that a json blob isn't especially well suited to a transcript format in this sense. Potentially we could have a "header" blob, and then the rest of the blobs just with the actual log information.

So it would look more like this (Copying from the above examples)

```json
{
    "transcriptHeader": "Windows PowerShell transcript start",
    "startTimeUtc": 20190123080857,
    "username": "BK-WINDOWS-V2-1\\petermounce",
    "runAsUser": "BK-WINDOWS-V2-1\\petermounce",
    "machine": "BK-WINDOWS-V2-1 (Microsoft Windows NT 10.0.14393.0)",
    "hostApplication": "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
    "processId": 7264,
    "psVersion": "5.1.14393.2636",
    "psEdition": "Desktop",
    "psCompatibleVersions": ["1.0", "2.0", "3.0", "4.0", "5.0", "5.1.14393.2636"],
    "buildVersion": "10.0.14393.2636",
    "clrVersion": "4.0.30319.42000",
    "wsManStackVersion": "3.0",
    "psRemotingProtocolVersion": "2.3",
    "serializationVersion": "1.1.0.1"
},
{ "output": "Processing a thing..." },
{ "verbose": "Performed <action> on <item>" },
{ "output": "Done." },
```

Certainly, this would make reading it a _little_ trickier, but still very much able to be processed automatically without much hassle. Just read the header blob and then read the subsequent lines.

**Comment by iSazonov** [comment]
If the goal is parsing, we need a _scheme_.

**Comment by iSazonov** [comment]
I'm not sure that traces are suitable for permanent use. These would be regular logs.

**Comment by petemounce** [comment]
I'm trying to illustrate what is normal, for me and others using centralised logs in a devops/SRE environment.

**Comment by lukexjeremy** [comment]
Interesting scenario @petemounce, one I too think PS ought to accommodate. As I see it, the discussion so far indicates that the `-LineTemplate` idea is a better fit than a fixed JSON format. As you pointed out originally, a template system allows the user to make whatever trade-off suits them.

Here are my initial thoughts on a template system.

`MshHostUserInterface.TranscribeResult` looks like the place to intercept the command's result text and instantiate the template before flushing to logs.

Code executing in that method would have access to a wealth information. As an implementer, I'd be asking what information should be exposed and how. Then that would determine how to validate a template supplied by the user.

**Comment by iSazonov** [comment]
We need discuss  logging vs tracing vs transcription to get right solution.
